### PR TITLE
Zoom to multiple selected layer(s) in layer tree

### PR DESCRIPTION
**Pull request by TurboGraphxBeige**
## Description

(Issue #40459)

This pull request implements the changes necessary to the "Zoom to Layer" function (available from the layer tree context menu and the View menu) so it can zoom to the extent of all the selected layers in the layer tree.

Instead of feeding currentLayer() into QgsLayerTreeViewDefaultActions::zoomToLayer(), it feeds selectedLayers(). zoomToLayers() is then called with a list of layers and loops through them to zoom to their combined extent, like the "Zoom to Group" function.

The menu item labels in the layer tree context menu and in the View menu are also changed to "Zoom to Layer(s)" to reflect the ability to zoom to one or more layers.

ezgif-7-91c5beb1d125

### CODE REVIEW

File: src/gui/layertree/qgslayertreeviewdefaultactions.cpp [hunk 1]
```diff
@@ -283,12 +283,10 @@ void QgsLayerTreeViewDefaultActions::showFeatureCount()
 
 void QgsLayerTreeViewDefaultActions::zoomToLayer( QgsMapCanvas *canvas )
```
Comment (m-kuhn) [inline src/gui/layertree/qgslayertreeviewdefaultactions.cpp:284]: Given the current name in the API, I think we'll need a new `zoomToLayers` method (and the current one can be deprecated).

**Comment by m-kuhn** [comment]
Looks useful and good :+1:

File: src/gui/layertree/qgslayertreeviewdefaultactions.cpp (hunk 1 above)
Comment (TurboGraphxBeige) [reply src/gui/layertree/qgslayertreeviewdefaultactions.cpp]: You mean like this commit? 

https://github.com/TurboGraphxBeige/QGIS/commit/4cb616df1077d988b8dec0be8710042bf5971339

Comment (TurboGraphxBeige) [reply src/gui/layertree/qgslayertreeviewdefaultactions.cpp:284]: For the sake of my new contributor's mind and soul...
You mean like this commit?

https://github.com/TurboGraphxBeige/QGIS/commit/e5ae5c47c91e2be90d0e48268a2b7c776e870853

Comment (nyalldawson) [reply src/gui/layertree/qgslayertreeviewdefaultactions.cpp:284]: @TurboGraphxBeige yep, looks perfect

Comment (TurboGraphxBeige) [reply src/gui/layertree/qgslayertreeviewdefaultactions.cpp:284]: Alright. Is it better if I move this commit into this PR or should I make a second PR?

Comment (nyalldawson) [reply src/gui/layertree/qgslayertreeviewdefaultactions.cpp:284]: Just cherry pick it into this branch please !

**Comment by m-kuhn** [comment]
Thanks for the update.
There might have been a misunderstanding regarding "deprecating". This means to leave the old version in place, parallel to the new version and it needs a `Q_DECL_DEPRECATED`; otherwise we will break the API and existing plugins.

**Comment by TurboGraphxBeige** [comment]
> 
> 
> Thanks for the update.
> There might have been a misunderstanding regarding "deprecating". This means to leave the old version in place, parallel to the new version and it needs a `Q_DECL_DEPRECATED`; otherwise we will break the API and existing plugins.

Oh ok!

So just to make sure, all the changes in https://github.com/TurboGraphxBeige/QGIS/commit/e5ae5c47c91e2be90d0e48268a2b7c776e870853 are correct except the fact that we need a new QgsLayerTreeViewDefaultActions::zoomToLayer function instead of just rename it?

**Comment by nyalldawson** [comment]
Ah sorry, my fault -- I didn't realise QgsLayerTreeViewDefaultActions was part of stable API

**Comment by TurboGraphxBeige** [comment]
I think (hope) I got this right with the last commit. Let me know!

File: src/gui/layertree/qgslayertreeviewdefaultactions.cpp [hunk 2]
```diff
@@ -324,6 +343,15 @@ void QgsLayerTreeViewDefaultActions::zoomToLayer()
   QApplication::restoreOverrideCursor();
 }
 
+void QgsLayerTreeViewDefaultActions::zoomToLayers()
+{
+  QAction *s = qobject_cast<QAction *>( sender() );
+  QgsMapCanvas *canvas = reinterpret_cast<QgsMapCanvas *>( s->data().value<void *>() );
```
Comment (m-kuhn) [inline src/gui/layertree/qgslayertreeviewdefaultactions.cpp:349]:
```suggestion
  QgsMapCanvas *canvas = s->data().value<QgsMapCanvas *>();
```

File: src/gui/layertree/qgslayertreeviewdefaultactions.h [hunk 3]
```diff
@@ -57,7 +57,12 @@ class GUI_EXPORT QgsLayerTreeViewDefaultActions : public QObject
     //! Action to check a group and all its parents
     QAction *actionCheckAndAllParents( QObject *parent = nullptr );
 
-    QAction *actionZoomToLayer( QgsMapCanvas *canvas, QObject *parent = nullptr ) SIP_FACTORY;
+    /**
+     * \deprecated since QGIS 3.18, use actionZoomToLayers()
+     */
+    Q_DECL_DEPRECATED QAction *actionZoomToLayer( QgsMapCanvas *canvas, QObject *parent = nullptr ) SIP_FACTORY;
+
+    QAction *actionZoomToLayers( QgsMapCanvas *canvas, QObject *parent = nullptr ) SIP_FACTORY;
```
Comment (m-kuhn) [inline src/gui/layertree/qgslayertreeviewdefaultactions.h:65]: Can you add a (short) api doc for this method?

File: src/gui/layertree/qgslayertreeviewdefaultactions.cpp [hunk 4]
```diff
@@ -93,7 +93,18 @@ QAction *QgsLayerTreeViewDefaultActions::actionZoomToLayer( QgsMapCanvas *canvas
...
   connect( a, &QAction::triggered, this, static_cast<void ( QgsLayerTreeViewDefaultActions::* )()>( &QgsLayerTreeViewDefaultActions::zoomToLayer ) );
+  Q_NOWARN_DEPRECATED_POP
+  return a;
+}
+
+QAction *QgsLayerTreeViewDefaultActions::actionZoomToLayers( QgsMapCanvas *canvas, QObject *parent )
+{
+  QAction *a = new QAction( QgsApplication::getThemeIcon( QStringLiteral( "/mActionZoomToLayer.svg" ) ),
+                            tr( "&Zoom to Layer(s)" ), parent );
+  a->setData( QVariant::fromValue( reinterpret_cast<void *>( canvas ) ) );
```
Comment (m-kuhn) [inline src/gui/layertree/qgslayertreeviewdefaultactions.cpp:106]:
```suggestion
  a->setData( QVariant::fromValue( canvas ) );
```

File: src/gui/layertree/qgslayertreeviewdefaultactions.h [hunk 5]
```diff
@@ -96,7 +101,12 @@ class GUI_EXPORT QgsLayerTreeViewDefaultActions : public QObject
      */
     QAction *actionMutuallyExclusiveGroup( QObject *parent = nullptr ) SIP_FACTORY;
 
-    void zoomToLayer( QgsMapCanvas *canvas );
+    /**
+    * \deprecated since QGIS 3.18, use zoomToLayers()
+    */
+    Q_DECL_DEPRECATED void zoomToLayer( QgsMapCanvas *canvas );
+
+    void zoomToLayers( QgsMapCanvas *canvas );
```
Comment (m-kuhn) [inline src/gui/layertree/qgslayertreeviewdefaultactions.h:109]: docstring also here

**Review by m-kuhn — COMMENTED** [review]
Looks almost good now.

- There's a cleaner way than reintepret cast (I can see that you copied it)
- Add some api docs for the new methods

**Comment by m-kuhn** [comment]
Hi @TurboGraphxBeige 

I finally got back to look at this again.
One thing regarding public api still needs to be adjusted, the qgisinterface (with qgisapp) is also considered stable, so we need to keep the existing `zoomToLayer` method in there.

Regarding the test failures, they originate from some derived files not being up to date. if you run `./scripts/sipify_all.sh` from the source folder, they will be updated. For the future, if you setup your pre commit hook this will make git fix this for you in the background automatically.

**Comment by TurboGraphxBeige** [comment]
@m-kuhn Do you mean `actionZoomToLayer`?

https://qgis.org/pyqgis/3.16/gui/QgisInterface.html?highlight=zoomtolayer#qgis.gui.QgisInterface.actionZoomToLayer

**Comment by m-kuhn** [comment]
Yes exactly. Here it must be added, not renamed.

https://github.com/qgis/QGIS/pull/40460/files#diff-236c824e6450ade0b821aa1f87e474ffa572123b91666ab8cfaa639d37c9a1ecR479

**Comment by TurboGraphxBeige** [comment]
I think we're getting there. I re-added actionZoomToLayer but I just didn't actually "<addaction ...>" the action since it didn't make sense to have both of them exposed in the interface.

**Comment by m-kuhn** [comment]
Sounds good !
Did you have a look at the test results?

**Comment by TurboGraphxBeige** [comment]
> Sounds good !
> Did you have a look at the test results?

Yeah but there's still 4 tests failing (I added a pre-commit hook executing `sipify_all.sh`). Am I missing something?

**Comment by TurboGraphxBeige** [comment]
> Sounds good !
> Did you have a look at the test results?

I didn't pay much attention since it's building fine on my machine but I think I'll have to dig a little bit more looking into the details for each failing tests. 

There's the "Windows cross build with MXE" that fails because of a warning though. I am not sure if I should let this one go.

Thank you for you patience, I must say I didn't expect pulling that much energy from you for this PR. I am learning a lot in the process and hopefully the next ones will go much smoother.

**Comment by m-kuhn** [comment]
That's what it says:

```
../src/app/qgsmapcanvasdockwidget.cpp:98:12: error: use of undeclared identifier 'mActionZoomToLayers'; did you mean 'mActionZoomToLayer'?
  connect( mActionZoomToLayers, &QAction::triggered, mMapCanvas, [ = ] { QgisApp::instance()->layerTreeView()->defaultActions()->zoomToLayers( mMapCanvas ); } );
           ^~~~~~~~~~~~~~~~~~~
           mActionZoomToLayer
src/ui/ui_qgsmapcanvasdockwidgetbase.h:28:14: note: 'mActionZoomToLayer' declared here
    QAction *mActionZoomToLayer;
             ^
```

**Comment by TurboGraphxBeige** [comment]
I fixed most of the failing checks except "Build with Qt 5.14 / build (pull_request)" and "continuous-integration/travis-ci/pr". I am unsure what to do with them.

**Comment by m-kuhn** [comment]
I think you can remove the old one from mapcanvasdockwidget.

**Comment by TurboGraphxBeige** [comment]
All checks passed, yay!

**Comment by nyalldawson** [comment]
Thank you @TurboGraphxBeige !

**Pull request merged**